zebra-chain/src/transaction.rs

```diff
@@ -318,14 +318,39 @@ impl Transaction {
     }
 
     /// Get this transaction's lock time.
-    pub fn lock_time(&self) -> LockTime {
-        match self {
-            Transaction::V1 { lock_time, .. } => *lock_time,
-            Transaction::V2 { lock_time, .. } => *lock_time,
-            Transaction::V3 { lock_time, .. } => *lock_time,
-            Transaction::V4 { lock_time, .. } => *lock_time,
-            Transaction::V5 { lock_time, .. } => *lock_time,
+    pub fn lock_time(&self) -> Option<LockTime> {
+        let lock_time = match self {
+            Transaction::V1 { lock_time, .. }
+            | Transaction::V2 { lock_time, .. }
+            | Transaction::V3 { lock_time, .. }
+            | Transaction::V4 { lock_time, .. }
+            | Transaction::V5 { lock_time, .. } => *lock_time,
+        };
+
+        // A lock time with a block height of zero means in practice that the lock time is
+        // disabled because it was effectively unlocked at the genesis block.
+        if lock_time == LockTime::unlocked() {
+            return None;
         }
+
+        // Consensus rule:
+        //
+        // > The transaction must be finalized: either its locktime must be in the past (or less
+        // > than or equal to the current block height), or all of its sequence numbers must be
+        // > 0xffffffff.
+        //
+        // This means that the lock time is only enabled if at least one input sequence number is
+        // not `u32::MAX`. The code below will make the method return `None` if such sequence
+        // number isn't found.
```

Comment:
Required Consensus rule clarifications:

1. Bitcoin doesn't allow transactions with no transparent inputs. This is Zcash-specific extension to transactions, so we need to explicitly follow `zcashd` here.

2. Coinbase inputs have a sequence number.

Optional:

3. Let's add a summary sentence without any negations.

```suggestion
        // In `zcashd`, this rule applies to both coinbase and prevout input sequence numbers.
        //
        // Unlike Bitcoin, Zcash allows transactions with no transparent inputs. These transactions only have shielded inputs.
        // Surprisingly, the `zcashd` implementation ignores the lock time in these transactions.
        // `zcashd` only checks the lock time when it finds a transparent input sequence number that is
        // not `u32::MAX`.
```